specify amount in /more

kennyrkun opened the pull request:
This pull request closes #2342.

This pull request allows users to specify an amount when using the /more command. The command still make sure to fulfill the following:
- if the user does not have essentials.oversizedstacks, the command will make sure not to give users stacks sized over the maximum limit for that item.
- the command will make sure not to give users stacks above Essentials:OversizedStackSize

This is a simple improvement for /more, not necessarily necessary. Do with it what you will.

edit: I have not tested and cannot test this, though I am confident it is functional.

kennyrkun committed:
implement /more tweak as described in EssentialsX/Essentials#2342

kennyrkun committed:
make it work for the 1.13 branch

i don't really want to do it this way, but it's all i know how to do.

if I'm correct then it will work regardless as we're only merging these files into the 1.13 branch, and not the 2.x branch.

Ichbinjoe commented on Essentials/src/com/earth2me/essentials/commands/Commandmore.java:
```diff
@@ -24,14 +24,34 @@ public void run(final Server server, final User user, final String commandLabel,
+
+            try {
+                newAmount += Integer.parseInt(ChatColor.stripColor(args[1]));
+            } catch(Exception e) {
```
Should be NumberFormatException

Ichbinjoe commented on Essentials/src/com/earth2me/essentials/commands/Commandmore.java:
```diff
@@ -24,14 +24,34 @@ public void run(final Server server, final User user, final String commandLabel,
+            int newAmount = stack.getAmount();
+
+            try {
+                newAmount += Integer.parseInt(ChatColor.stripColor(args[1]));
```
Why would we need to strip color from this argument?

kennyrkun replied:
I don't know exactly how `parseInt` functions but I assumed it would fail if it received something with a formatting code. We strip the colour from the argument so that it will work even if a formatting code is present.

Ichbinjoe replied:
Why would a client be sending us an argument for a number with color? Its a good thing if it fails - what the client has sent us is not a number - it is a formatting code plus a number.

kennyrkun replied:
You're right, I suppose it would be a better idea to reject all but raw numbers.

kennyrkun replied:
~~Is this PR gonna have to add a custom entry to the messages files or is there something else we can use to send the error?~~

edit: my bad, I realise that it already has an error which works.

Ichbinjoe commented on Essentials/src/com/earth2me/essentials/commands/Commandmore.java:
```diff
@@ -24,14 +24,34 @@ public void run(final Server server, final User user, final String commandLabel,
+            }
+
+            if (!user.isAuthorized("essentials.oversizedstacks")) {
+                newAmount = stack.getMaxStackSize();
```
Whenever the user does not have the essentials.oversizedstacks permission, newAmount will always be the max stack size, regardless of what the user entered.

kennyrkun replied:
I don't understand how that happens, would you please explain it to me?

Ichbinjoe replied:
So its a logic error - lets say args[1] is '1', and I'm currently holding 3. Lets also say I'm holding grass, which has a stacksize of 64. Lets ALSO say I do not have essentials.oversizedstacks.

Given the above, by the time we get to line 37 `if (!user.isAuthorized("essentials.oversizedstacks")) {` `newAmount` will be equal to 4 - 1 + 3. However, since I don't have essentials.oversizedstacks, we will go into the if block, and `newAmount` will be overridden to be 64.

Instead of setting newAmount, we should enforcing the max stack size and erroring (as discussed below) if we exceed it.

Ichbinjoe commented on Essentials/src/com/earth2me/essentials/commands/Commandmore.java:
```diff
@@ -24,14 +24,34 @@ public void run(final Server server, final User user, final String commandLabel,
+            if (!user.isAuthorized("essentials.oversizedstacks")) {
+                newAmount = stack.getMaxStackSize();
+            } else {
+                if (newAmount > ess.getSettings().getOversizedStackSize()) {
```
Should we be capping this amount, or should we instead throw an error?

kennyrkun replied:
I think an error would be better here, as capping the amount might cause some confusion for users.

I can't seem to find a suitable string in messages.yml, so I think we're gonna have to add another one for this purpose. Unless there's another method, however, I'm new to the Essentials codebase so I don't know of any.

Ichbinjoe replied:
Nope, thats the correct way to do it. I'm also fairly certain you need to dig up translations for all of the translation variations.

kennyrkun committed:
remove strip color from input

we don't want to fix the number for the user, because their input is invalid.

updated spacing, and changing exception to NumberFormatException

kennyrkun committed:
fix missing parameter space on line 41

the build should pass now

mdcfe commented:
This was closed by mistake due to deleting the 1.13 branch.

kennyrkun commented:
So, I don't really know what I'm doing here.

After the 1.13 branch was merged into the 2.x, I think my branch is now out of date. How do I update it and keep the changes but not flood this pull request with unnecessary commits?

Also, for completion, this pull request still requires the following:
- [ ] an error for stacks that would exceed the max stack limit
  - [ ] A new string in the messages
    - [ ] update every translation we have

Ichbinjoe commented:
@kennyrkun you want to do something called a rebase + force push. https://git-scm.com/book/en/v2/Git-Branching-Rebasing

From what I remember, the checklist you present is accurate to what needs changed in this PR.

mdcfe commented on Essentials/src/com/earth2me/essentials/commands/Commandmore.java:
```diff
@@ -24,14 +24,31 @@ public void run(final Server server, final User user, final String commandLabel,
+            } catch(NumberFormatException e) {
+                throw new Exception(tl("numberRequired"));
+            }
+            if (!user.isAuthorized("essentials.oversizedstacks")) {
```
You should check whether they're exceeding the max stack size before limiting it.
```java
            if (newAmount > stack.getMaxStackSize() && !user.isAuthorized("essentials.oversizedstacks")) {
                newAmount = stack.getMaxStackSize();
            } else {
                if (newAmount > ess.getSettings().getOversizedStackSize()) {
                    newAmount = ess.getSettings().getOversizedStackSize();
                }
            }
```

mdcfe reviewed (changes requested):
I don't think throwing an error when the player gives an excessively high number is necessary (`/more 999` would be a useful quick way to fill a stack), but there is a logic error here that needs fixing.

Ichbinjoe commented:
Looks like it has been quite a while since this PR has seen any love from @kennyrkun... I'm going to close. Feel free to re-PR.
